### ISSUE TITLE
Request Workerman support

### ISSUE THREAD

**Issue by Yurunsoft** [laravel/octane#160]
Workerman is an asynchronous event-driven PHP framework with high performance to build fast and scalable network applications. Workerman supports HTTP, Websocket, SSL and other custom protocols. Workerman supports event extension.

<https://github.com/walkor/workerman>

**Comment by barryvdh** [comment]
What is the benefit over Workerman over RoadRunner and Swoole? No binary/extension required? It does support PSR-7 so probably wouldn't be too hard to create; https://github.com/walkor/Workerman/issues/51#issuecomment-684685099
But there is also amphp, php-pm etc

**Comment by szepeviktor** [comment]
https://github.com/php-pm/php-pm is a very nice project.
Will Octane support it?

**Comment by barryvdh** [comment]
Maybe instead of trying to support them all, we could see if it's possible to add servers with a 3rd party package? I think the most part doesn't need to live in Octane directly, as long as you can add the server somewhere.

**Comment by barryvdh** [comment]
For reference, I create initial Workerman support. As you can see, I didn't need to change any of the core functions so far, only the ServiceProvider to add bindings (but you can do that 3rd party). Furthermore to actually start it with the `octane:start` instead of `octane:workerman` command, we would need to add it as an available server somewhere.
https://github.com/barryvdh/octane/commit/eac68e9835a096ef4661cb6a5a7803893023dacb

(Not this isn't really tested, but it starts Workerman with the `octane:workerman --host=0.0.0.0 --port=1234 --workers=4` parameters correctly, it serves Laravel and it shuts down the server. I didn't really get it to hot reload yet though (although I think I'm sending the right command). Bit hard to debug further though because I have never used Workerman.

I can PR this but I'll leave it to the Octane team to see what their intention is for other providers.

**Comment by barryvdh** [comment]
If you abstract it a bit more, at first we could consider moving the start command to the ServerInspector class, just as stop/reload. The state-file is mostly the same for servers like Workerman/Roadrunner/php-pm that have a master process (I guess at least) so maybe there can be 1 common ServerStateFile instead of per driver. 
If the ServerInspector handles the starting and also the output formatting, there could be a general `octane:start-server --server=roadrunner` which just uses the registered ServerInspectors to start.

But ofcourse this is just thinking out load. Not sure which drivers there could be more, besides php-pm and Workerman now.

**Comment by taylorotwell** [comment]
No plans right now. It would primarily depend on how much code it is to maintain and what advantages it would offer against existing drivers.

**Comment by taylorotwell** [comment]
I think @barryvdh's suggestion of allowing this to be extendable third-party is a good idea.

**Comment by Salmondx** [comment]
Sorry for "necroposting", but have you guys had time to think how to integrate these third party octane plugins?

Been playing with Workerman for a couple weeks and so far it have been great. Main selling point for it is that it's codebase is pretty "simple", its pure php (just a composer package based on pcntl and optional event extension) and it doesn't require installing any third party extensions that you have to trust (like swoole for example) or require any third party web servers like roadrunner. Another side bonus is that its easy to setup xdebug with it.

It is hard to say about its performance (don't have any production experience with workerman) but based on artificial benchmarks (including some tests with PDO and multiple queries) it performs pretty well: https://www.techempower.com/benchmarks/#section=data-r21&hw=ph&test=fortune
Code for such tests is simple (https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/PHP/workerman/app.php) and it's hard to predict how it will work on real life apps.

It doesn't provide any cool features like swoole's coroutines but looks like PHP is slowly heading towards native concurrency anyways (recent Fibers RFC as I understand is a foundation for stuff that they plan to do in future).

TLDR; nice option for octane to have, single composer dependency without the need of any external extensions or web servers.

@nunomaduro sorry for a ping but it is such a nice proposal from barryvdh that imho worth reevaluation :smile:

**Comment by mouyong** [comment]
> What is the benefit over Workerman over RoadRunner and Swoole? No binary/extension required? It does support PSR-7 so probably wouldn't be too hard to create; [walkor/workerman#51 (comment)](https://github.com/walkor/workerman/issues/51#issuecomment-684685099) But there is also amphp, php-pm etc

@taylorotwell After integrating Workerman, it can make the development of some functions very simple, such as Workerman's IoT development, Workerman's WebSocket, Workerman's cluster deployment Gateway, Workerman's customization of any protocol docking, and so on.

And currently, third-party Laravel Octane Worker is in a state of being archived in the warehouse due to technical and energy limitations. As one of them, I really hope that Laravel can integrate the Octane: Worker feature officially.
The reference code for the warehouse is: https://github.com/JieAnthony/laravel-octane-workerman and https://github.com/mouyong/laravel-octane-workerman

**Comment by mouyong** [comment]
> Been playing with Workerman for a couple weeks and so far it have been great. Main selling point for it is that it's codebase is pretty "simple", its pure php (just a composer package based on pcntl and optional event extension) and it doesn't require installing any third party extensions that you have to trust (like swoole for example) or require any third party web servers like roadrunner. Another side bonus is that its easy to setup xdebug with it.

@taylorotwell And I think this point is very correct. It is very simple to use and also very easy to extend custom protocols. The official documentation of Workerman is very easy to understand.

**Comment by mouyong** [comment]
@taylorotwell And if we want to deploy Laravel on Kubernetes, using `octane:start --server=workerman` will be easier to accomplish than using other service providers such as `octane:start --server=swoole`
